templates/README.md

```diff
@@ -0,0 +1,221 @@
+
+The UI can operate on top of template parameters and/or the Template.VM.spec DOM directly.
+
+## Example file
+
+### template.yaml
+
+```YAML
+# vim: sts=2 sw=2 et
+apiVersion: v1
+kind: Template # Openshift kind of template or something "better"
+metadata:
+  name: windows-10
+  annotations:
+    description: "Basic windows 10 template"
+    
+    # This set of annotations is meant as a hint only, it is
+    # not going to be processed by the stock openshift templating
+    # mechanism. CNV UI can and should use it though.
+    # The information encoded here informs the tooling (UI)
+    # about names of devices that are to be used as templates
+    # for adding additional disk, networks, volumes and such
+    defaults.template.cnv.io/disk: default-disk
+    defaults.template.cnv.io/volume: default-volume
+    defaults.template.cnv.io/network: default-network
+
+    # Replacements useful for finishing the transformation
+    # This is an extension we could use for specifying common
+    # fields for all disks/nics. Requires extra implementation
+    # that will do the find and replace using a syntax documented
+    # here: https://kubernetes.io/docs/reference/kubectl/jsonpath/
+    template.cnv.io/replace:
+    - path: "/VM/disks/*/type"
+      value: virtio
+
+    # Another extension useful for CNV aware tooling that allows
+    # expressing additional validation rules for this template
+    template.cnv.io/validate:
+    - path: "/VM/spec/domain/guest/memory"
+      min: 3
+      max: 4
+
+    # Another extension for hinting at which elements should be
+    # considered editable
+    template.cnv.io/editable:
+      - "/VM/spec/domain/guest/memory"
+      - "/VM/spec/domain/guest/memory"
+
+  labels:
+    # The UI can show all possible template.cnv.io/* values in a nice way
+    # and let the user filter down the available templates to the one
+    # the user actually wants:
+    # A single selected template only means no conflicts and no smart
+    # merging code. This has to be done using labels to allow efficient
+    # searching.
+    # The format has the following meaning:
+    # {os,flavor,size}.template.cnv.io/{value}: {version} or just true
+    os.template.cnv.io/windows2k12r2: 201809 
+    os.template.cnv.io/windows8: 201809 
+    os.template.cnv.io/windows7: 201809 
+    flavor.template.cnv.io/minimal: 201809
+    # io.cnv.vm.size not specified means all?
+
+# Parameters must come from a subset of well known names
+# so the UI can properly work with those.
+parameters:
+- description: VM name
+  from: '[A-Za-z0-9]{1,16}'
+  generate: expression
+  name: VM_NAME
+- description: Memory size
+  from: '[A-Z0-9]{8}'
+  generate: expression
+  name: MEMORY_SIZE
+  value: 1Gi
+
+objects:
+# The full VM template with placeholders for either scalars like memory
+# or yaml structures like disks
+- apiVersion: v1
+  kind: VirtualMachine
```

Comment:
In this exampel teh VM is pretty small.

We either provide all the details using presets or directly within the template.

The mental model if we have, template, vm, vmi only is simpler than with presets.

Reply:
We want all VMs to be based on a template if created this way and we should provide a Blank template as well (possibly with just the OS bits). And I agree that the mental model of template -> vm -> vmi is simpler.